### ISSUE TITLE
Null pointer exception when using TypeScript 1.7.3

### ISSUE THREAD

**Issue by DatenMetzgerX** [palantir/tslint#832]
Typescript 1.7 has been released. The current version fails with the following error

```js
Module build failed: TypeError: Cannot read property 'kind' of undefined
  at isStatementBraced (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/rules/curlyRule.js:92:16)
  at CurlyWalker.visitDoStatement (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/rules/curlyRule.js:75:14)
  at CurlyWalker.SyntaxWalker.visitNode (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/walker/syntaxWalker.js:293:22)
  at /Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/walker/syntaxWalker.js:455:63
  at visitEachNode (/Users/micha/git/kw-advisory-gui/node_modules/typescript/lib/typescript.js:7421:30)
  at Object.forEachChild (/Users/micha/git/kw-advisory-gui/node_modules/typescript/lib/typescript.js:7578:24)
  at CurlyWalker.SyntaxWalker.walkChildren (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/walker/syntaxWalker.js:455:12)
  at CurlyWalker.SyntaxWalker.visitNode (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/walker/syntaxWalker.js:449:22)
  at CurlyWalker.SyntaxWalker.walk (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/walker/syntaxWalker.js:21:14)
  at Rule.AbstractRule.applyWithWalker (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/language/rule/abstractRule.js:33:16)
  at Rule.apply (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/rules/curlyRule.js:29:21)
  at Linter.lint (/Users/micha/git/kw-advisory-gui/node_modules/tslint/lib/tslint.js:41:37)
  at Object.lint (/Users/micha/git/kw-advisory-gui/node_modules/tslint-loader/index.js:55:23)
  at Object.module.exports (/Users/micha/git/kw-advisory-gui/node_modules/tslint-loader/index.js:139:8)
```

It works with tslint@next. But we can not use next, as a peerDependency locks down the version.

**Comment by blakeembrey** [comment]
:+1:

**Comment by SomeKittens** [comment]
Looks like `curlyRule.ts` is passing `node.statement` to `isStatementBraced` which is checking `node.kind`.  `kind` is a property of `node` but not `node.statement`.

Full stack trace:

```
node_modules/tslint/lib/rules/curlyRule.js:95
    return node.kind === 190;
               ^

TypeError: Cannot read property 'kind' of undefined
    at isStatementBraced (node_modules/tslint/lib/rules/curlyRule.js:95:16)
    at CurlyWalker.visitDoStatement (node_modules/tslint/lib/rules/curlyRule.js:77:14)
    at CurlyWalker.SyntaxWalker.visitNode (node_modules/tslint/lib/language/walker/syntaxWalker.js:293:22)
    at node_modules/tslint/lib/language/walker/syntaxWalker.js:455:63
    at visitEachNode (node_modules/typescript/lib/typescript.js:7421:30)
    at Object.forEachChild (node_modules/typescript/lib/typescript.js:7578:24)
    at CurlyWalker.SyntaxWalker.walkChildren (node_modules/tslint/lib/language/walker/syntaxWalker.js:455:12)
    at CurlyWalker.SyntaxWalker.visitNode (node_modules/tslint/lib/language/walker/syntaxWalker.js:449:22)
    at CurlyWalker.SyntaxWalker.walk (node_modules/tslint/lib/language/walker/syntaxWalker.js:21:14)
    at Rule.AbstractRule.applyWithWalker (node_modules/tslint/lib/language/rule/abstractRule.js:33:16)
```

**Comment by adidahiya** [comment]
that's unfortunate. we've been testing with the `next` version of the compiler, but looks like whatever released from TypeScript's `release-1.7` branch has some deviations from the dev releases on `next`. I'll take a look soon.

**Comment by adidahiya** [comment]
Looking more closely, this is simply fixed by compiling `tslint` with TS 1.7.3. I haven't looked deeply into why that fixes it, but in the mean time I'll release 3.1 to get around the issue.

**Comment by adidahiya** [comment]
re-opening until the release with the fix is published.

**Comment by adidahiya** [comment]
Ok, I've traced the error back to the const enum problem which was raised in October: https://github.com/palantir/tslint/pull/726#issuecomment-147883525

What happens is that if you attempt to use a version of `typescript` that is different from the version that `tslint` was compiled with, the linter ends up __visiting the wrong walker methods__ because of the switch statement in `syntaxWalker.ts` which relies on the `ts.SyntaxKind` const enum.

We shouldn't have version incompatibility problems like this going forward, though, since TS has deconst'd their const enums in `typescript.d.ts`: https://github.com/Microsoft/TypeScript/pull/5422.

For now, I have to bump our peer dependency to >= 1.7.3: https://github.com/palantir/tslint/pull/840

**Comment by adidahiya** [comment]
published v3.1.1 to NPM.

**Comment by DatenMetzgerX** [comment]
thx

**Comment by nchutchind** [comment]
I'm not sure what the procedure for this is, but I'm getting this error with TSLint@3.13.0 and TypeScript@2.0.0. If I revert TypeScript to 1.8.10, the lint works fine. If I remove the "one-line" rule or set it to false, the lint works fine otherwise with Typescript@2.0.0.

Here is the offending rule:
```
"one-line": [
    true,
    "check-open-brace",
    "check-catch",
    "check-else",
    "check-finally",
    "check-whitespace"
],
```

And stacktrace:

```
if (body.kind === ts.SyntaxKind.ModuleBlock) {
        ^

TypeError: Cannot read property 'kind' of undefined
    at OneLineWalker.visitModuleDeclaration (c:\dev\sandbox\node_modules\tslint\lib\rules\oneLineRule.js:163:17)
    at OneLineWalker.SyntaxWalker.visitNode (c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:366:22)
    at c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:459:63
    at visitEachNode (c:\dev\sandbox\node_modules\typescript\lib\typescript.js:7959:30)
    at Object.forEachChild (c:\dev\sandbox\node_modules\typescript\lib\typescript.js:8117:24)
    at OneLineWalker.SyntaxWalker.walkChildren (c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:459:12)
    at OneLineWalker.SyntaxWalker.visitSourceFile (c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:190:14)
    at OneLineWalker.SyntaxWalker.visitNode (c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:414:22)
    at OneLineWalker.SyntaxWalker.walk (c:\dev\sandbox\node_modules\tslint\lib\language\walker\syntaxWalker.js:7:14)
    at Rule.AbstractRule.applyWithWalker (c:\dev\sandbox\node_modules\tslint\lib\language\rule\abstractRule.js:19:16)
    at Rule.apply (c:\dev\sandbox\node_modules\tslint\lib\rules\oneLineRule.js:21:21)
```

**Comment by adidahiya** [comment]
@nchutchind try using `tslint@next`. `tslint@latest` (3.13.0) has not yet been compiled with TS 2.0.

**Comment by nchutchind** [comment]
@adidahiya 
I switched to tslint@next (3.14.0-dev.0) and typescript@2.0.0 and re-enabled the "one-line" rule. The same versions are installed globally and in my project.

I still get the same error and stacktrace as before.

It's not a huge deal for me, because I can just comment out the rule, but I thought you should know.

**Comment by itrethan** [comment]
@adidahiya Doesnt works for me either.

**Comment by JKillian** [comment]
@nchutchind or @itrethan  Could you post a small code snippet that causes TSLint to throw the error you mentioned? If you could do it in a new issue it would be great (easier to keep track of and find)

**Comment by nchutchind** [comment]
@JKillian Interesting... I created a new project with the same versions of tslint and typescript (and same tslint.json), but the error is not happening on my simple script that I'm linting. There must be something specific with my other, larger project that is causing the error to come up. I tested with typescript@2.0.0 and tslint at "3.13.0" and "next" and it worked every time.

I'll create a new issue if I can ever pinpoint the cause in the other project, but for now I'll leave it alone.

**Comment by nchutchind** [comment]
@JKillian @adidahiya Thank you for your help, both of you.

**Comment by JKillian** [comment]
No problem! Keep us updated if you reproduce it minimally, it's definitely something that would be a high priority to fix

**Comment by nchutchind** [comment]
@JKillian Actually... I tried one last thing and added a console.log to the oneLineRule.js file in node_modules so I could see the node it was complaining about. Turns out it's a weird line from a custom type definitions file where the author was trying to... suppress missing module errors, I think? The project is a modified version of the [AngularClass Webpack Starter](https://github.com/AngularClass/angular2-webpack-starter). The specific file containing the offending line is [custom-typings.d.ts](https://github.com/AngularClass/angular2-webpack-starter/blob/master/src/custom-typings.d.ts).

    declare module "*";

Is this worth putting into a new issue, or is it behavior that shouldn't be happening in the first place, since it looks like that line of code is possibly experimental (from what I can see in [TypeScript Issue#6615](https://github.com/Microsoft/TypeScript/issues/6615))?
